README
Bio-Grep version 0.10.5
=======================

Bio-Grep is a collection of Perl modules for searching in DNA and protein
sequences. It supports different back-ends, most importantly some (enhanced) 
suffix array implementations. Currently, there is no suffix array tool that
works in all scenarios (for example whole genome, protein and RNA data). 
Bio::Grep provides a common API to the most popular tools. This way, you 
can easily switch or combine tools.


DEPENDENCIES

This module requires these other modules, libraries and programs:


  Bioperl 1.4 http://bioperl.org/Core/Latest/index.shtml
    (You will need "Core" and "Run") Be sure that you have all
    required modules installed. You don't need to worry about this if 
    you follow all installation steps in the Bioperl Documentation 
    (for example you will need some XML modules. Bundle::BioPerl 
    will install them).
    Works with Bioperl 1.5.2.
    
  EMBOSS (ftp://emboss.open-bio.org/pub/EMBOSS/EMBOSS-latest.tar.gz)


  The back-ends:
  
  You need at least one of them. Vmatch is always a good choice. Agrep is
  the best choice if you allow many mismatches in short sequences, if you
  want to search in Fasta files with relatively short sequences (e.g
  transcript databases) and if you are only interested in which sequences
  the approximate match was found. Its performance is in this case 
  amazing. If you want the exact positions, choose vmatch. If you want 
  nice alignments, choose vmatch too (EMBOSS can automatically align the 
  sequence and the query in the agrep back-end, but then vmatch is faster). 
  Filters require exact positions, so you can't use them with agrep. 
  This may change in future version or not.
  GUUGle is the best choice if you have RNA queries (counts GU as no mismatch)
  and if you are interested in only exact matches.

  RE (perldoc perlre) for Perl regular expressions. A slow sliding window
      algorithm that searches all sequences for the specified regular
      expression. Compared to suffix arrays this is really slow for (nearly)
      perfect matches and large databases but requires no additional software.
      Currently it stores all hits of one sequence in memory. So this could be
      a problem for Fasta files with huge sequences (For example whole
      chromosomes as one sequence).

  Vmatch (http://vmatch.de/) for the Vmatch back-end. Commercial software.
    The Vmatch tests assume that vmatch is in your path (You
    can later specify a path to vmatch that is not in your path. 
    The tests will fail but the module should work if the 
    specified path to vmatch is correct.). 
 
  Agrep (http://www.tgries.de/agrep/) for the Agrep back-end. There
    are packages for some Linux distributions available (Debian:
    apt-get install agrep). Fink has some packages for Mac OS X.
    Ebuilds for Gentoo are available, too. As for Vmatch, Agrep 
    tests assume that agrep is in your path. 
    The TRE-agrep (http://laurikari.net/tre/) is much slower but has
    more features and less limitations. 

  GUUGle (http://bibiserv.techfak.uni-bielefeld.de/guugle/)
    A suffix array implementation for RNA sequences. Only allows search
    for exact matches. It is very memory efficent and needs no precalculated
    suffix arrays. Open Source.

  ...and other Perl modules. You will get a warning about missing
  modules when you run the make command. A lot of dependencies, we know,
  but most of them are standard software in bioinformatics. So please
  check if some of them are already installed on your workstation.


INSTALLATION

To install this module type the following (AFTER the installation of
the software in the "Dependecies"-section):

    perl Build.PL
    ./Build
    ./Build test
    ./Build install

Alternatively, to install it "the old way", you can use the following
commands:

   perl Makefile.PL
   make
   make test 
   make install


DOCUMENTATION


1. Tutorials
------------

bgrep is an example implementation. The source code is well documented, so
maybe it is a good starting point.

A not yet comprehensive coobook is available in perldoc Bio::Grep::Cookbook.
Please contribute recipes if you can!


2. Performance
--------------

2.1 Vmatch


*  Try $sbe->settings->showdesc(200) if you don't need upstream or downstream
   regions. This makes the parser get all data direcly out of vmatch output.
   Otherwise the parser will call vsubseqselect for every search result.

*  Try $sbe->settings->online(1) if you allow many mismatches.


3. FAQ
------

- Is it possible to get the coordinates of the hit out of the alignment?
  Yes. $res->alignment->get_seq_by_pos(1)->...

  see perldoc Bio::SimpleAlign


BUGS

Please report any bugs, recipes for the cookbook or feature requests to
C<bug-bio-grep@rt.cpan.org>, or through the web interface at
L<http://rt.cpan.org>. 


COPYRIGHT AND LICENCE

Copyright (C) 2007-2008 by M. Riester.

Based on Weigel::Search v0.13, Copyright (C) 2005-2006 by Max Planck 
Institute for Developmental Biology, Tuebingen.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.4 or,
at your option, any later version of Perl 5 you may have available.



Changes
Revision history for Perl extension Bio::Grep.

0.10.5  Sat Jul  26 13:00:00 2008
    - META.yml broken 
    - small cleanup in Vmatch back-end
    - RE backend: switched to $+ and $- to retrieve the positions. 

0.10.4  Fri Jul  25 17:53:00 2008
    - New Perl::Critic passes
    - added Bio:Perl dependencies
    - Bugfix: -blanks in dirnames allowed (thanks Jamie)

0.10.3  Sun Apr  17 12:01:00 2008
    - code cleanup:
        * perlcritic severity 1
    - Bugfix: RNA databases allowed
        
0.10.2  Sun Sep  23 16:08:00 2007
    - code cleanup:
        * renamed test data files
        * perlcritic
        * renamed FilterI->reset to FilterI->reset_filter
          because reset is a buildin function
        * renamed SearchSettings->set to SearchSettings->set_attributes
          cause set is an ambigious name
    - more tests:
        * features.updownstream (more or less moved from guugle/RE specific
          tests)
    - New Features:
        * upstream / downstream in SearchResult
        * reverse_complement in SearchResult
    - Bugfixes:
        * Vmatch: $res->sequence undef when available downstream region
                  to small

0.10.1  Mon Sep  10 11:54:00 2007
    - Minor POD improvements
    - Benchmarks: 
       * generated with Template::Toolkit
       * added vmatch -online
       * ./Build dist asks whether benchmarks should be created

0.10.0  Tue Aug  21 17:00:00 2007
    - POD improvements:
        * methods now grouped in CONSTRUCTOR, ACCESSORS/MUTATORS, PACKAGE
          METHODS, INTERNAL METHODS, ABSTRACT METHODS
        * notes now under IMPORTANT NOTES, not DESCRIPTION  
    - code cleanup:
        * standard testfile "header" now in BioGrepTest, requires ToolSet
    - more tests:
        * features.alignments
    - New Features:
        * Bio::Seq objects as query 
    - Bugfixes:
        - bgrep not installed witch Build.PL
        - Added Bioperl dependency (ticket #28881, thanks siddhartha)
        - $res->query, $res->alignments not equal in back-ends
          now, '(reverse complement)' is added to $query->desc  

0.9.2  Fri Aug  17 12:33:00 2007
    - POD improvements
    - huge performance improvements (RE, Agrep)
    - support for TRE agrep
    - skip EMBOSS alignment calculation when hits can't have gaps
    - idsonly option in bgrep
    - better exception texts
    - sequence description in RE results
    - benchmarks
    
0.9.1  Tue Aug  15 16:10:00 2007
    - POD improvements
    - current directory default datapath (thanks Rick)
    - removed index from *.map files
      (you need to rebuild Agrep/RE dbs!)
    - add -pl option in generate_database (prefix_length)
    - skipping databases generation when database with that
      name already exists 
    - code cleanup: 
         - SearchResult->new wants a hashref as argument
         - RE.pm
         - perltidy
    - tests check warnings
    - more tests (especially GUUGle and RE)
    - maxhits for RE
    - Bugfixes
        - RE pretended to have mismatches feature
    
0.9.0  Mon Aug  08 18:05:00 2007
    - POD improvements
    - generate_database now wants a hashref as argument,
      symlinks fasta file per default, not copies (closes ticket #28691,
      thanks Rick),
      in theory, not only fasta files supported
      Vmatch: non-verbose call of mkvtree. Closes ticket #28720 (thanks Rick)   
    - $sbe->results now throws an deprecated warning 
    - removed HyPa back-end. seems to be a dead project and vmatch covered
      most features
    - Bugfixes
        - tests did not rm temporary directories

0.8.5  Mon Aug  06 13:41:00 2007
    - POD improvements 
    - Bugfixes:
       - SYNOPSIS test failed when Vmatch binary not in path. Closes ticket
         #28656 (thanks Malcolm)
       - arrays not possible with SearchSettings->set (and therefore also not in
         $sbe->search)

0.8.4  Fri Aug  03 14:37:00 2007
    - POD improvements 
    - code cleanup
    - RE back-end much faster

0.8.3  Thu Aug  02 20:00:00 2007
    - POD improvements 
    - dynamically loaded back-ends
    - check arguments of $sbe->get_sequences
    - more tests
    - Bio::Grep now returns a back-end object.
    - Bugfixes:
        - GUUGle back-end confused when sequence description missing
          with up-/downstream regions

0.8.2  Tue Jul  31 16:11:00 2007
    - POD improvements 
    - more tests
    - code cleanup (Bio::Grep::Root)
    
0.8.1  Tue Jul  31 16:11:00 2007
    - POD improvements 
    - code cleanup
    - Cookbook
    - removed directory Container (unnecessary and confusing)
    - Error messages more consistent
    - more tests
    
0.8.0  Mon Jul  30 19:20:00 2007
    - POD improvements
    - direct_and_rev_com
    - RE back-end
    - use Fatal qw(open close), removed own exception handling
    - replaced global dirhandle in BackendI with local dirhandle
    - more tests
    - renamed generate_database_out_of_fastafile() to
      generate_database()
    - renamed Backends/ to Backend  
    - Bugfixes:
        - GUUGle back-end confused when sequence description missing

0.7.0  Tue Jul  24 16:55:00 2007
    - POD improvements:
       * added section exception handling in Bio::Grep
       * removed exception handling in the back-ends
         SYNOPSIS
       * polishing 
    - Bugfixes:
       * Agrep back-end: called two times 

0.6.0  Tue Mar  27 14:09:10 2007
    - POD improvements
    - BackendI code cleanups 
    - renamed feature QUERYFILE to QUERY_FILE
    - did not throw an exception when query and query_file
      are set
    - more tests
    - removed $res->get_gene_id and $res->alignment_to_csv
    
0.5.0  Sat Mar  23 17:34:20 2007
    - POD improvements
    - different values for upstream and downstream in GUUGle
    - did not throw an exception when alphabet file was missing
    - database generation in bgrep broken
    - more tests

0.4.0  Fri Mar  23 15:59:00 2007
    - hxdrop, exdrop flags for vmatch
    - more tests
    - POD improvements

lib/Bio/Grep/Backend/GUUGle.pm
#############################################################################
#   $Author: markus $
#     $Date: 2008-07-26 18:37:36 +0200 (Sat, 26 Jul 2008) $
# $Revision: 813 $
#############################################################################

package Bio::Grep::Backend::GUUGle;

use strict;
use warnings;

use version; our $VERSION = qv('0.10.5');

use Fatal qw(open close);
use English qw( -no_match_vars );

use Bio::Grep::SearchResult;
use Bio::Grep::Backend::BackendI;

use base 'Bio::Grep::Backend::BackendI';

use List::Util qw(max);


sub new {
    my $self = shift;
    $self = $self->SUPER::new;
    my %all_features = $self->features;
    delete $all_features{EVALUE};
    delete $all_features{PERCENT_IDENTITY};
    delete $all_features{MISMATCHES};
    delete $all_features{DELETIONS};
    delete $all_features{INSERTIONS};
    delete $all_features{EDITDISTANCE};
    delete $all_features{ONLINE};
    delete $all_features{COMPLETE};
    delete $all_features{SHOWDESC};
    delete $all_features{QSPEEDUP};
    delete $all_features{PROTEINS};
    delete $all_features{HXDROP};
    delete $all_features{EXDROP};
    delete $all_features{NATIVE_D_A_REV_COM};
    $self->settings->gumismatches(0);
    $self->features(%all_features);
    return $self;
}

sub search {
    my ( $self, $arg_ref ) = @_;
    my $s = $self->settings;
    $self->_check_search_settings($arg_ref);

    my ( $query, $query_file, $tmp_query_file )
        = $self->_create_tmp_query_file();

    # now generate the command string
    my $eflag = q{};
    if ( $s->upstream > 0 || $s->downstream > 0 ) {
        $eflag = ' -e ' . max( $s->upstream, $s->downstream ) . q{ };
    }

    if ( $s->gumismatches > 0 && !$s->direct_and_rev_com ) {
        $self->warn( "GUUGle counts GU always as no mismatch.\n"
                . 'Set gumismatches(0) to hide this warning.' );
    }
    if ( !$s->reverse_complement && $s->query_file ) {
        $self->throw(
            -class => 'Bio::Root::BadParameter',
            -text => 'GUUGle searches only for the reverse ' . 'complement. ',
            -value => $s->reverse_complement,
        );
    }

    my $lflag = q{};
    if ( $s->maxhits_isset ) {
        $lflag = ' -l ' . $s->maxhits . q{ };
    }

    #set query_length automatically
    my $auto_query_length = 0;
    if ( !defined $s->query_length ) {
        if ($query) {
            $s->query_length( length $query );
            $auto_query_length = 1;
        }
        else {
            $self->throw(
                -class => 'Bio::Root::BadParameter',
                -text  => 'query_length not set. See -d flag in the '
                    . 'GUUGle documentation. ',
            );
        }
    }

    my $command = $self->_cat_path_filename( $s->execpath, 'guugle' ) . $eflag
        . $lflag . ' -d '
        . $s->query_length . q{ }
        . $self->_cat_path_filename( $s->datapath, $s->database ) . q{ }
        . $tmp_query_file;

    if ( $ENV{BIOGREPDEBUG} ) {
        warn $command . "\n";
    }

    my $cmd_ok = $self->_execute_command($command);

    if ( !$cmd_ok ) {
        $self->throw(
            -class => 'Bio::Root::SystemException',
            -text  => "GUUGle call failed. Command was:\n\t$command"
        );
    }

    $self->_skip_header();
    my %query_desc_lookup;
    foreach my $query_seq ( @{ $self->_query_seqs } ) {

        # simulate how this sequence would look in guugle output
        my $query_desc = $query_seq->id;
        if ( $query_seq->desc && length($query_desc) > 0 ) {
            $query_desc .= q{ } . $query_seq->desc;
        }
        $query_desc_lookup{$query_desc} = $query_seq;
    }
    $self->{_mapping} = \%query_desc_lookup;
    $self->_prepare_results;
    if ($auto_query_length) {
        $self->settings->query_length_reset;
    }
    return 1;
}

sub get_databases {
    my $self = shift;
    return $self->_get_databases('.al1');
}

sub generate_database {
    my ( $self, @args ) = @_;

    my %args = $self->_prepare_generate_database(@args);

    if ( defined $args{skip} ) {
        return 0;
    }

    $self->_create_index_and_alphabet_file( $args{filename} );
    return 1;
}

sub _skip_header {
    my ($self)      = @_;
    my $FH          = $self->_output_fh;
    my $blank_lines = 0;
    while ( my $line = <$FH> ) {
        chomp $line;

        # skip everything before first line
        if ( $line =~ m{\A \s* \z}xms ) {
            $blank_lines++;
            return if $blank_lines == 2;
        }
    }
    return;
}

sub _rnas_match {
    my ( $self, $s1, $s2 ) = @_;
    return 1 if $s1 eq $s2;
    return 0 if length $s1 != length $s2;

    # now check for wobble pairs
    for my $i ( 0 .. ( length($s1) - 1 ) ) {
        my $base_a = substr $s1, $i, 1;
        my $base_b = substr $s2, $i, 1;
        if (   $base_a ne $base_b
            && join( q{}, $base_a, $base_b ) ne 'uc'
            && join( q{}, $base_a, $base_b ) ne 'ga' )
        {
            return 0;
        }
    }
    return 1;
}

sub _get_subject_id_and_desc {
    my ( $self, $text ) = @_;
    if ( $text =~ m{(.*?)\s(.*)}xms ) {
        return ( $1, $2 );
    }
    else {
        return ( $text, q{} );
    }
}

sub _parse_next_res {
    my $self       = shift;
    my $s          = $self->settings;
    my @query_seqs = $self->_query_seqs;

    # temp variables. for parsing only
    my ($subject_id,  $subject_desc, $subject_pos,
        $subject_seq, $query_desc,   $query_pos,
        $query_seq,   $matchlength,  $upstream
    ) = ( 0, q{}, 0, 0, q{}, 0, 0, 0, 0 );

    my ( $up_seq, $down_seq ) = ( q{}, q{} );

    my $FH = $self->_output_fh;
LINE:
    while ( my $line = <$FH> ) {
        chomp $line;

        if ($line =~ m{\A
                MatchLength:\s(\d+)\s
                "(.*?)"
                \sat\s(\d+)\s
                vs\.\s
                "(.*?)"\s
                at\s(\d+)
             }xms
            )
        {
            (   $matchlength, $subject_id, $subject_pos,
                $query_desc,  $query_pos
            ) = ( $1, $2, $3, $4, $5 );
            ( $subject_id, $subject_desc )
                = $self->_get_subject_id_and_desc($subject_id);

            next LINE;
        }

        if ($line =~ m{\A
                >(.*?)
                _at_(\d+)
                _with_
                (.*?)
                _at_(\d+)
             }xms
            )
        {    # -e mode

            ( $subject_id, $subject_pos, $query_desc, $query_pos, )
                = ( $1, $2, $3, $4, );
            ( $subject_id, $subject_desc )
                = $self->_get_subject_id_and_desc($subject_id);
            next LINE;
        }

        if ( $line =~ m{ \A 5 (.*) 3 \z}xms ) {
            $subject_seq = $1;
            next LINE;
        }

        if ( $line =~ m{ \A 3 (.*) 5 \z}xms ) {
            $query_seq = $1;
            next LINE;
        }

        if ( $line =~ m{\A Maximum\snumber}xms ) {
            last LINE;
        }

        # find the query that belongs to the match
        my $query = $self->{_mapping}->{$query_desc};

        # already reverse, so just the complement here:
        $query_seq =~ tr[UTGCAutgca][AACGUaacgu];

        # -e mode
        if ( $line =~ m{ \A [gucaGUCA]+ \z }xms ) {

            # find subject in string. a little bit complicated because
            # don't know if the upstream/downstream region is as large
            # as we request
            my $qrc = lc $query->revcom->seq;
            $qrc =~ s/t/u/gxms;
            my $ql = $s->query_length || length $query->seq;
        SUBSTRING:
            for my $length ( reverse $ql .. length $query->seq ) {
            STARTPOSITION:
                for my $start (
                    reverse 0 .. max( $s->upstream, $s->downstream ) )
                {
                    my $query_start
                        = length( $query->seq ) - $query_pos + 1 - $length;
                    my $qs = substr $qrc,  $query_start, $length;
                    my $ss = substr $line, $start,       $length;

                #warn "L:$length S:$start QS:$query_start $qrc $qs $line $ss";
                    next STARTPOSITION if !( $self->_rnas_match( $ss, $qs ) );
                    $subject_pos = $start;
                    $subject_seq = $ss;
                    $query_seq   = $qs;
                    $matchlength = $length;
                    $upstream    = $start;
                    $up_seq      = substr $line, 0, $upstream;
                    $down_seq    = substr $line, $upstream + $length;

                    # now check if regions are larger than
                    # expected. can happen when up/down differ
                    if ( length($up_seq) > $s->upstream ) {
                        $up_seq = substr $up_seq,
                            length($up_seq) - $s->upstream;
                        $upstream = length $up_seq;
                    }
                    if ( length($down_seq) > $s->downstream ) {
                        $down_seq = substr $down_seq, 0, $s->downstream;
                    }
                    last SUBSTRING;
                }
            }
        }
        my $fasta = Bio::Seq->new(
            -id   => $subject_id,
            -seq  => $up_seq . $subject_seq . $down_seq,
            -desc => $subject_desc,
        );
        my $tmp_aln = new Bio::SimpleAlign( -source => 'Bio::Grep' );
        $tmp_aln->add_seq(
            Bio::LocatableSeq->new(
                -id    => 'Subject',
                -seq   => $subject_seq,
                -start => $subject_pos,
                -end   => $subject_pos + length($subject_seq) - 1,
            )
        );
        $tmp_aln->add_seq(
            Bio::LocatableSeq->new(
                -id    => $query->id,
                -desc  => $query->desc,
                -seq   => $query_seq,
                -start => $query_pos,
                -end   => $query_pos + length($query_seq) - 1,
            )
        );
        my $res = $self->_filter_result(
            Bio::Grep::SearchResult->new(
                {   sequence    => $fasta,
                    begin       => $upstream,
                    end         => $upstream + $matchlength,
                    alignment   => $tmp_aln,
                    sequence_id => $fasta->id,
                    remark      => q{},
                    query       => $query,
                }
            )
        );
        if ($res) {
            return $res;
        }
    }
    close $FH;

    #$self->_delete_output();
    return 0;
}

sub get_sequences {
    my ( $self, $seqid ) = @_;
    return $self->_get_sequences_from_bio_index($seqid);
}

sub available_sort_modes {
    my ($self) = @_;

    # get sort modes from superclass
    return ( $self->SUPER::available_sort_modes() );
}
1;    # Magic true value required at end of module
__END__


=head1 NAME

Bio::Grep::Backend::GUUGle - GUUGle back-end  


=head1 SYNOPSIS

  use Bio::Grep;
  
  my $sbe = Bio::Grep->new('GUUGle');
  
  # generate a GUUGle Bio::Grep database. you have to do this only once.
  # GUUGle does not create a persistent index right now.
  # This function generates an fast index for $sbe->get_sequences
  # and files with a description and the alphabet (only DNA/RNA allowed)
  $sbe->generate_database({ 
    file        => 'ATH1.cdna', 
    description => 'AGI Transcripts',
    datapath    => 'data',
  });
 
  # search on both strands (GU allowed) 
  # retrieve up- and downstream regions of size 30
  $sbe->search({
    query   => 'AGAGCCCT',
    direct_and_rev_com => 1, 
    upstream           => 30,
    downstream         => 30,
    gumismatches       => 0,
    database           => 'ATH1.cdna',
  });

  my @internal_ids;

  # output all informations we have!
  while ( my $res = $sbe->next_res ) {
     print $res->sequence->id . "\n";
     print $res->mark_subject_uppercase() . "\n";
     print $res->alignment_string() . "\n\n";
     push @internal_ids, $res->sequence_id;
  }
  
  # get the complete sequences as Bio::SeqIO object
  my $seq_io = $sbe->get_sequences(\@internal_ids);
  
  # search for targets (GU allowed)
  $sbe->search({
    query   => 'GAGCCCTTGGGGGGG',
    reverse_complement => 1, 
    gumismatches       => 0,
  });

=head1 DESCRIPTION

B<Bio::Grep::Backend::GUUGle> searches for a query in a C<GUUGle> suffix 
array. 


=head1 METHODS

See L<Bio::Grep::Backend::BackendI> for inherited methods. 

=head2 CONSTRUCTOR

=over 

=item C<Bio::Grep::Backend::GUUGle-E<gt>new()>

This method constructs a C<GUUGle> back-end object and should not used 
directly. Rather, a back-end should be constructed by the main class 
L<Bio::Grep>:

  my $sbe = Bio::Grep->new('GUUGle');

=back

=head2 PACKAGE METHODS

=over

=item C<$sbe-E<gt>available_sort_modes()>

Returns all available sort modes as hash. keys are sort modes, values a short
description.

   $sbe->sort('ga');

Available sortmodes in C<GUUGle>:

=over

            ga  : 'ascending order of dG'
            gd  : 'descending order of dG'

=back

Note that 'ga' and 'gd' require that search results have dG set. 
L<Bio::Grep::RNA> ships with filters for free energy calculation. Also note that
these two sort options require that we load all results in memory.

=back

=head1 IMPORTANT NOTES

=over

=item C<reverse_complement>

C<GUUGle> always searches for the reverse complement. If you specify
a query file, C<Bio::Grep> throws an exception if C<reverse_complement> is 
not set.

=item C<mismatches>

C<GUUGle> only allows search for exact matches. It counts GU as no
mismatch.

=item C<maxhits>

When "maxhits" is defined, this back-end returns the I<maxhits> first
hits.

=back

=head1 DIAGNOSTICS

See L<Bio::Grep::Backend::BackendI> for other diagnostics. 

=over

=item C<GUUGle call failed. Command was: ...> 

It was not possible to run C<GUUGle> in function search(). Check the search
settings. If you want to reproduce the system() call, you can set the 
environment variable C<BIOGREPDEBUG>. If this variable is set, then the temporary
files won't get deleted. C<Bio::Root::SystemException>.

=item C<GUUGle searches only for the reverse complement.> 

You have specified a query file and C<reverse_complement> is not set. 
C<Bio::Root::BadParameter>.

=item C<query_length not set. See -d flag in the...>

You have specified a query file and forgot to set C<query_length>.
C<Bio::Root::BadParameter>.

=back

=head1 SEE ALSO

L<Bio::Grep::Backend::BackendI>
L<Bio::Grep::SearchSettings>
L<Bio::Seq>


=head1 AUTHOR

Markus Riester, E<lt>mriester@gmx.deE<gt>

=head1 LICENCE AND COPYRIGHT

Copyright (C) 2007-2008 by M. Riester. 

This module is free software; you can redistribute it and/or
modify it under the same terms as Perl itself.


=head1 DISCLAIMER OF WARRANTY

BECAUSE THIS SOFTWARE IS LICENSED FREE OF CHARGE, THERE IS NO WARRANTY
FOR THE SOFTWARE, TO THE EXTENT PERMITTED BY APPLICABLE LAW. EXCEPT WHEN
OTHERWISE STATED IN WRITING THE COPYRIGHT HOLDERS AND/OR OTHER PARTIES
PROVIDE THE SOFTWARE "AS IS" WITHOUT WARRANTY OF ANY KIND, EITHER
EXPRESSED OR IMPLIED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE. THE
ENTIRE RISK AS TO THE QUALITY AND PERFORMANCE OF THE SOFTWARE IS WITH
YOU. SHOULD THE SOFTWARE PROVE DEFECTIVE, YOU ASSUME THE COST OF ALL
NECESSARY SERVICING, REPAIR, OR CORRECTION.

IN NO EVENT UNLESS REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING
WILL ANY COPYRIGHT HOLDER, OR ANY OTHER PARTY WHO MAY MODIFY AND/OR
REDISTRIBUTE THE SOFTWARE AS PERMITTED BY THE ABOVE LICENCE, BE
LIABLE TO YOU FOR DAMAGES, INCLUDING ANY GENERAL, SPECIAL, INCIDENTAL,
OR CONSEQUENTIAL DAMAGES ARISING OUT OF THE USE OR INABILITY TO USE
THE SOFTWARE (INCLUDING BUT NOT LIMITED TO LOSS OF DATA OR DATA BEING
RENDERED INACCURATE OR LOSSES SUSTAINED BY YOU OR THIRD PARTIES OR A
FAILURE OF THE SOFTWARE TO OPERATE WITH ANY OTHER SOFTWARE), EVEN IF
SUCH HOLDER OR OTHER PARTY HAS BEEN ADVISED OF THE POSSIBILITY OF
SUCH DAMAGES.

=cut

# vim: ft=perl sw=4 ts=4 expandtab
